src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -294,6 +294,63 @@ public void ConnectHeaders_SupplyHeadersThenSupplyMore_Both_Tip_PeerId_Maps_Shou
+        /// 3 peers should supply headers - one of the peers creates a fork.
+        /// Make sure that ChainedBlock is not created more than once for the same header.
+        /// Check that next pointers were created correctly.
+        /// </summary>
+        [Fact]
+        public void ConnectHeaders_HeadersFromTwoPeersWithFork_ShouldCreateBlocksForNewHeaders()
+        {
+            TestContext testContext = new TestContextBuilder().Build();
+            ChainedHeaderTree chainedHeaderTree = testContext.ChainedHeaderTree;
+            ChainedHeader chainTip = testContext.ExtendAChain(7);
+            chainedHeaderTree.Initialize(chainTip, true);
+            testContext.ChainStateMock.Setup(s => s.ConsensusTip).Returns(chainTip);
+
+            // Peer 1: 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a
+            Assert.True(chainedHeaderTree.GetChainedHeadersByHash().Count == 8);
+
+            // Peer 2: 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a
+            List<BlockHeader> listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 7);
+            chainedHeaderTree.ConnectNewHeaders(2, listOfExistingHeaders);
+            Assert.True(chainedHeaderTree.GetChainedHeadersByHash().Count == 8);
+
+            // Peer 3: 1a - 2a - 3a - 4a - 5b - 6b - 7b - 8b
+            var forkedBlockHeader = chainTip.Previous.Previous.Previous.Previous;
+            ChainedHeader peerThreeTip = testContext.ExtendAChain(4, forkedBlockHeader);
+            List<BlockHeader> listOfNewAndExistingHeaders = testContext.ChainedHeaderToList(peerThreeTip, 6);   // includes common blocks
+            chainedHeaderTree.ConnectNewHeaders(3, listOfNewAndExistingHeaders);
+
+            // CHT should contain 12 headers
+            Assert.True(chainedHeaderTree.GetChainedHeadersByHash().Count == 12);
+
+            var chainedHeadersByHash = chainedHeaderTree.GetChainedHeadersByHash();
+
+            Assert.True(chainTip.Height==7);
+            for (int height = chainTip.Height; height > 0; height--)
+            {
+                var chainedHeadersByHashAtHeight = chainedHeadersByHash.Where(x => x.Value.Height == height).ToArray();
+
+                int blocksAtHeight = chainedHeadersByHashAtHeight.Count();
+                if (height < 4) Assert.True(blocksAtHeight == 1);
+                if (height > 4) Assert.True(blocksAtHeight == 2);
+
+                // Each should have 1 Next pointer
+                if (height < 3 || (height > 3 && height < 7))
+                {
+                    Assert.True(chainedHeadersByHashAtHeight.All(x => x.Value.Next.Count == 1));
+                }
+
+                // Except for 8a and 8b which contain none
+                if (height == 7) Assert.True(chainedHeadersByHashAtHeight.All(x => x.Value.Next.Count == 0));
+
+                // And 4a which contains 2
+                if (height == 3) Assert.True(chainedHeadersByHashAtHeight.All(x => x.Value.Next.Count == 2));
+            }
```

Comment:
This is still not what we were looking for. The main problem of this approach is that you only check the number of next pointers, but you don't check whether they are correctly set to the next header. This is why I suggested you should start at the tip and use previous pointers to traverse the chain down to the Genesis. You just have two tips of two different chains so that would mean just doing two loops.

I do not insist on the traversal using previous pointers, it just seems to me that it would be much easier code to read because of very simple logic that does not mix the two chains together. However, checking the next pointers properly should be done regardless of which way to traverse you choose.

Reply:
Rewritten accordingly.